### PR TITLE
[worker, training_utils] fix: Engine Metric Aggregation

### PR DESCRIPTION
**Pull request by JacobHelwig**
### What does this PR do?

Because some metrics are scaled by global_bsz/global_tokens in `workers.utils.losses.ppo_loss`, the mean in `reduce_metrics` adds an extra scaling of the metric by the number of gradient accumulation steps (see examples in Test sec): 
https://github.com/volcengine/verl/blob/c191c5eb5c9499dca6666a52bc5f360bfd4bbf4f/verl/utils/metric/utils.py#L53

Aggregation of the `loss` metric handles this by taking sum: https://github.com/volcengine/verl/blob/c191c5eb5c9499dca6666a52bc5f360bfd4bbf4f/verl/workers/engine_workers.py#L143-L144
Depending on how metrics are handled in `workers.utils.losses.ppo_loss`, it may not be correct to aggregate all of them using sum (as in #4785). For example, `actor/pg_loss` and `actor/kl_loss` are scaled by global batch sizes/ token counts, and should be aggregated using sum, while the `pg_metrics` `pg_clipfrac`, `ppo_kl`, and `pg_clipfrac_lower` are scaled by local token counts and should be aggregated using mean.

This PR introduces a metric management class to allow flexibility in deciding the aggregation type on a per-metric basis.

### Test

This test demonstrates the scaling of metrics with the number of gradient accumulation steps, as well as how this is resolved on this branch. The command for running is below.

image

```bash
gsm8k_train_path=$DATA_DIR/gsm8k/train.parquet
gsm8k_test_path=$DATA_DIR/gsm8k/test.parquet

train_files="['$gsm8k_train_path']"
test_files="['$gsm8k_test_path']"

ppo_micro_batch_size_per_gpu=2
ppo_micro_batch_size_per_gpu=8

branch=main
branch=fixEngineMetrics

python3 -m verl.trainer.main_ppo \
    algorithm.adv_estimator=grpo \
    data.dataloader_num_workers=0 \
    data.return_full_prompt=True \
    data.train_files="$train_files" \
    data.val_files="$test_files" \
    data.train_batch_size=8 \
    data.max_prompt_length=512 \
    data.max_response_length=1024 \
    data.filter_overlong_prompts=True \
    data.truncation='error' \
    actor_rollout_ref.model.path=Qwen/Qwen3-0.6B \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.actor.ppo_mini_batch_size=8 \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=$ppo_micro_batch_size_per_gpu \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.fsdp_config.param_offload=True \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=True \
    actor_rollout_ref.rollout.tensor_model_parallel_size=1 \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=8 \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=8 \
    actor_rollout_ref.rollout.name=vllm \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.6 \
    actor_rollout_ref.rollout.n=5 \
    trainer.logger='["console","wandb"]' \
    trainer.project_name='fixEngineMetrics' \
    trainer.experiment_name="$branch/ppo_micro_batch_size_per_gpu$ppo_micro_batch_size_per_gpu" \
    trainer.n_gpus_per_node=2 \
    trainer.nnodes=1 \
    trainer.save_freq=400 \
    trainer.test_freq=40 \
    trainer.use_legacy_worker_impl=disable \
    trainer.total_epochs=2 \
    trainer.total_training_steps=10 \
    trainer.resume_mode=disable \
    actor_rollout_ref.actor.use_torch_compile=False \
    actor_rollout_ref.actor.fsdp_config.use_torch_compile=False \
    trainer.val_before_train=False \
    actor_rollout_ref.rollout.enforce_eager=True \
    actor_rollout_ref.ref.fsdp_config.use_torch_compile=False
```

### Design & Code Changes

Adds a `Metric` class which tracks metric values and aggregation type.

### CODE REVIEW

**Comment by wuxibin89** [comment]
@JacobHelwig Please format code according to: https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting

File: verl/utils/py_functional.py [hunk 1]
```diff
@@ -196,7 +198,7 @@ def append_to_dict(data: dict, new_data: dict, prefix: str = ""):
     for key, val in new_data.items():
         new_key = f"{prefix}{key}" if not key.startswith(prefix) else key
         if new_key not in data:
-            data[new_key] = []
+            data[new_key] = Metric.init_list(val) if isinstance(val, Metric) else []
```
[the remark this answers is not in the document]
Comment (wuxibin89) [reply verl/utils/py_functional.py]: @JacobHelwig Please address gemini's review, `init_list` is not a classmethod.

File: verl/utils/metric/utils.py [hunk 2]
```diff
@@ -45,10 +47,84 @@ def reduce_metrics(metrics: dict[str, list[Any]]) -> dict[str, Any]:
...
+    SUM = "sum"
+    MIN = "min"
+    MAX = "max"
+
+
+NumericType = int, float, torch.Tensor
+Numeric = int | float | torch.Tensor
+
+
+class Metric:
```
Comment (wuxibin89) [inline verl/utils/metric/utils.py:72]: Could you also add a unit test for this.

Comment (JacobHelwig) [reply verl/utils/metric/utils.py:72]: Added!

Comment (wuxibin89) [inline verl/utils/metric/utils.py:72]: @JacobHelwig Docstring Coverage ci failed, please add doc string for Metric.

Comment (JacobHelwig) [reply verl/utils/metric/utils.py:72]: Added!

**Comment by pengwu22** [comment]
Hi @JacobHelwig, thanks for catching the metrics issue!

I think the fix needs one more adjustment. While the summation over micro-batches is correct, the reduction across data-parallel ranks should use **mean** (not sum) in `reduce_metrics` at `ray_trainer.py`, since the actual loss values are already correctly scaled by `dp_size` in `agg_loss`.

**Recommended approach:**
1. Keep `Metric.chain` in `train_mini_batch` to sum over micro-batches, OR explicitly aggregate metrics there without the `Metric` class
2. Use **mean** reduction in `reduce_metrics` at `ray_trainer.py`

This ensures the final metric correctly represents the global per-token average without being inflated by `dp_size`.

**Comment by JacobHelwig** [comment]
Nice catch, @pengwu22. Opened #5203 to resolve this.
